sklego/pipeline.py

```diff
@@ -0,0 +1,151 @@
+    '''
+    logger = logging.getLogger(__name__)
+    step_result, step = output
+    logger.info(f'[{step}] shape={step_result.shape} time={execution_time}')
+
+
+def _log_wrapper(func, log_callback=_default_log_callback):
+    '''Function wrapper to log information after the function is called, about
+    the output and the execution time.
+
+    Parameters
+    ----------
+    func : function
+        The function to be wrapped with a log statement.
+    log_callback : function, optional
+        The log callback which is called after `func` is called. Defaults to
+        :func:_default_log_callback. Note, this function should expect the
+        same arguments as the default.
+
+    Returns
+    ----------
+    function : The function wrapped with a log callback.
+    '''
+    def _(*args, **kwargs):
+        start_time = dt.datetime.now()
+        output = func(*args, **kwargs)
+        execution_time = dt.datetime.now() - start_time
+        log_callback(output, execution_time)
+        return output
+    return _
+
+
+def _cache_with_function_log_statement(log_callback=_default_log_callback):
+    '''Wraps the `func` with :func:_log_wrapper before passing it to
+    :method:_cache.
+
+    Parameters
+    ----------
+    log_callback : function, optional
+        The log callback function. Defaults to :func:_default_log_callback.
+
+    Returns
+    ----------
+    function : The cache where its function is wrapped with a log statement.
+    '''
+    def _(
+            self,
+            func=None,
+            *args,
+            **kwargs):
+        if callable(func):
+            func = _log_wrapper(func, log_callback=log_callback)
+        return self._cache(func=func, *args, **kwargs)
+    return _
+
+
+class Pipeline(Pipeline):
+    '''A pipeline that has a log statement in between each step, useful for
+    debugging.
+
+    Parameters
+    ----------
+    log_callback : function, optional
+        The callback function that logs information in between each
+        intermediate step. Defaults to None. If set to `'default'`,
+        :func:_default_log_callback is used.
+
+        See :func:_default_log_callback for an example.
+
+    See :class:sklearn.pipeline.PipeLine for all other parameters.
+
+    Note
+    ----------
+    This implementation is a hack on the original sklearn Pipeline. It aims to
+    have the same behaviour as the original sklearn Pipeline, while changing
+    minimal amount of code.
+
+    The log statement is added by overwriting the cache method of the memory,
+    such that the function called in the cache is wrapped with a functions that
+    calls the log callback function (`log_callback`).
+
+    This hack breaks will break when:
+    1) The SKlearn pipeline initialization function is changed.
+    2) The memory is used differently in the fit.
+    3) The :class:joblib.memory.Memory changes behaviour of the cache function.
+    4) The :class:joblib.memory.Memory starts using a `_cache` method.
```

Comment:
Two things. 

1. Have you checked if `make docs` renders this docstring appropriately? You do not need to send doc changes but because you're not using the `:param :` syntax this might be something to confirm. 
2. Is it possible to add an example to this docstring as well?

Reply:
Both are done.